labgrid/driver/shelldriver.py

```diff
@@ -71,39 +91,151 @@ def on_activate(self):
+
+        Args:
+            command (str): the command to run in background
+
+        Returns:
+            process handle (str)
+        """
+        return self._run_as_background_process(command)
+
+    def _run_as_background_process(self, command):
+        if self.background_process and self.background_process.running:
+            raise ExecutionError('Another background process is running')
+
         self._check_prompt()
         marker = gen_marker()
         # hide marker from expect
         cmp_command = '''MARKER='{}''{}' run {}'''.format(
-            marker[:4], marker[4:], shlex.quote(cmd)
+            marker[:4], marker[4:], shlex.quote(command)
         )
         self.console.sendline(cmp_command)
-        _, _, match, _ = self.console.expect(r'{marker}(.*){marker}\s+(\d+)\s+{prompt}'.format(
-            marker=marker, prompt=self.prompt
-        ), timeout=timeout)
+        self.console.expect(r'{marker}'.format(marker=marker), timeout=1)
+
+        self.background_process = ShellProcess(command, marker)
+
+        # provide ShellProcess object to satisfy the BackgroundProcessProtocol requirements and for
+        # step context
+        return self.background_process
+
+    @Driver.check_active
+    @step(args=['process', 'timeout'], result=True)
+    def read_from_background_process(self, process, *, timeout=0):
+        """
+        Read stdout/stderr of background process until timeout. For timeout=0 (default) the call
+        won't block and will return stdout/stderr until current EOF. For timeout=None the call
+        will block until the process has terminated. Returns a tuple (stdout, stderr).
+        Raises ExecutionError if process is not known.
+
+        Args:
+            process (str): process handle
+            timeout (int or None): will block until timeout is exceeded or process terminates,
+                                   timeout=0 returns immediately (default), timeout=None blocks
+                                   until process terminates
+
+        Returns:
+            (stdout (str), stderr (str))
+        """
+        return self._read_from_background_process(process, timeout=timeout)
+
+    def _read_from_background_process(self, process, *, timeout=0):
+        if process != self.background_process:
+            raise ExecutionError('Unknown process handle')
+
+        # skip reading if process terminated
+        if not self.background_process.running:
+            return '', ''
+
+        expect = [r'(.*){marker}\s+(\d+)\s+{prompt}'.format(
+            marker=self.background_process.marker,
+            prompt=self.prompt)]
+        if not timeout:
+            expect.append(TIMEOUT)
+
+        index, before, match, _ = self.console.expect(expect, timeout=timeout)
+
+        if index == 0:
+            self.background_process.exitcode = int(match.group(2))
+            stdout = match.group(1)
+        elif index == 1:
+            stdout = self.console._expect.read(size=len(before))
```

Comment:
Is there a better way to retrieve output until EOF without using the private `_expect.read()` here?